letmeind/src/protocol.rs

```diff
@@ -217,14 +237,26 @@ impl<'a, C: ConnectionOps> Protocol<'a, C> {
                     Ok(fw) => fw,
                 };
 
-                // Send an open-port request to letmeinfwd.
+                // Send an open-port or close-port request to letmeinfwd based on the operation type.
                 assert_eq!(self.auth_state, AuthState::ChallengeResponseAuth);
-                if let Err(e) = fw
-                    .open_port(self.conn.peer_addr().ip(), port_type, *port)
-                    .await
-                {
-                    let _ = self.send_go_away().await;
-                    return Err(err!("letmeinfwd firewall open: {e}"));
+                if operation == Operation::Close {
+                    // Close port operation
+                    if let Err(e) = fw
+                        .close_port(self.conn.peer_addr().ip(), port_type, *port)
+                        .await
+                    {
+                        let _ = self.send_go_away().await;
+                        return Err(err!("letmeinfwd firewall close: {e}"));
+                    }
+                } else {
```

Comment:
Please explicitly check for `else if operation == Operation::Knock` here for defensive coding.
The else branch could be made `unreachable!()`.